rcode-reader/src/main/java/eu/rssw/pct/PLReader.java

```diff
@@ -0,0 +1,174 @@
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package eu.rssw.pct;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.RandomAccessFile;
+
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.charset.Charset;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Class for reading and extracting contents of a Progress Library file.
+ * 
+ * @author <a href="mailto:g.querret+PCT@gmail.com">Gilles QUERRET</a>
+ */
+public class PLReader {
+  private static final int MAGIC_V11 = 0xd70b;
+  private static final int ENCODING_OFFSET = 0x02;
+  private static final int ENCODING_SIZE = 20;
+  private static final int FILE_LIST_OFFSET_V11 = 0x22;
+
+  private File pl;
+  private List<FileEntry> files = null;
+
+  public PLReader(File file) {
+    String name = file.getPath();
+    SecurityManager sm = System.getSecurityManager();
+    if (sm != null) {
+      sm.checkRead(name);
+    }
+
+    this.pl = file;
+  }
+
+  /**
+   * Returns entries contained in this procedure library
+   * 
+   * @throws RuntimeException If file is not a valid procedure library
+   */
+  public List<FileEntry> getFileList() {
+    if (this.files == null)
+      readFileList();
+    return files;
+  }
+
+  public FileEntry getEntry(String name) {
+    for (FileEntry entry : getFileList()) {
+      if (entry.getFileName().equals(name))
+        return entry;
+    }
+
+    return null;
+  }
+
+  private void readFileList() {
+    try (RandomAccessFile raf = new RandomAccessFile(pl, "r")) {
+      FileChannel fc = raf.getChannel();
+      ByteBuffer magic = ByteBuffer.allocate(2);
+      fc.read(magic);
+      if ((magic.getShort(0) & 0xffff) != MAGIC_V11)
+        throw new RuntimeException("Not a valid PL file");
+
+      Charset charset = getCharset(fc);
+      int offset = getTOCOffset(fc);
+      files = new ArrayList<>();
+      FileEntry fe = null;
+      while ((fe = readEntry(fc, offset, charset)) != null) {
+        if (fe.isValid())
+          files.add(fe);
+        offset += fe.getTocSize();
+      }
+    } catch (IOException caught) {
+      throw new RuntimeException(caught);
+    }
+  }
+
+  public InputStream getInputStream(FileEntry fe) throws IOException {
+    ByteBuffer bb = null;
+    try (RandomAccessFile raf = new RandomAccessFile(pl, "r")) {
+      FileChannel fc = raf.getChannel();
+      bb = ByteBuffer.allocate(fe.getSize());
+      fc.read(bb, fe.getOffset());
+    }
+
+    return new ByteArrayInputStream(bb.array());
+  }
+
+  private Charset getCharset(FileChannel fc) throws IOException {
+    ByteBuffer bEncoding = ByteBuffer.allocate(ENCODING_SIZE);
+    if (fc.read(bEncoding, ENCODING_OFFSET) != ENCODING_SIZE)
+      throw new RuntimeException("Invalid PL file");
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS00112)